### PR TITLE
Addresses #4183 and #4231, provide OS SDK/API support for the EnergyPlus "CoilSystem:IntegratedHeatPump:AirSource" and "Coil:WaterHeating:AirToWaterHeatPump:VariableSpeed" objects

### LINKED ISSUE #4183: Provide OS SDK/API support for the EnergyPlus "CoilSystem:IntegratedHeatPump:AirSource" object
Enhancement Request
--------------
Provide OS SDK/API support for the EnergyPlus "CoilSystem:IntegratedHeatPump:AirSource" object 

## Detailed Description
User interest in using OpenStudio to evaluate the performance of (both packaged and engineered systems) HVAC/DHW systems that combine ccASHP (cold climate heat pumps) performance with integrated hot water generation (and storage) by users may be relatively low today but has very high rate of change given interest in building de-carbonization and electrificatio.

While OS users do have access to this EnergyPlus object via an EnergyPlus Measure, the requirement to add the EnergyPlus measure at the 'tail end' of a measure based workflow combined with the requirement to provide connections to several other supporting OS objects (heating coil, cooling coil and water heating coil) can make this approach fragile and difficult to generalize, as needed to support for larger scale analysis workflows. 

## Possible Implementation
To the degree possible, a 1:1 mapping of the EnergyPlus object attributes would appear to be sufficient and a lowest effort method for implementation.

### LINKED ISSUE #4231: Provide OS SDK/API support for the EnergyPlus "Coil:WaterHeating:AirToWaterHeatPump:VariableSpeed" object
Enhancement Request
--------------

### PR DESCRIPTION
**Pull request by joseph-robertson**
Pull request overview
---------------------

 - Fixes #4183
 - Fixes #4231
 - Related to https://github.com/NREL/EnergyPlus/issues/8611
   - Autosizing the cooling coil (default ctor behavior) leads to error (crash)
   - Hardsizing the cooling coil misses this bug
   - This should be merged now because the E+ issue is independently under investigation and any (advanced) users of this object would effectively assist in debugging the object

Please read [OpenStudio Pull Requests](https://github.com/NREL/OpenStudio/wiki/OpenStudio-Pull-Requests) to better understand the OpenStudio Pull Request protocol.

### Pull Request Author

Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.

 - [ ] Model API Changes / Additions
 - [ ] Any new or modified fields have been implemented in the EnergyPlus ForwardTranslator (and ReverseTranslator as appropriate)
 - [ ] Model API methods are tested (in `src/model/test`)
 - [ ] EnergyPlus ForwardTranslator Tests (in `src/energyplus/Test`)
 - [ ] If a new object or method, added a test in NREL/OpenStudio-resources: Add Link
 - [ ] If needed, added VersionTranslation rules for the objects (`src/osversion/VersionTranslator.cpp`)
 - [ ] Checked behavior in OpenStudioApplication, adjusted policies as needed (`src/openstudio_lib/library/OpenStudioPolicy.xml`)
 - [ ] Verified that C# bindings built fine on Windows, partial classes used as needed, etc.
 - [ ] All new and existing tests passes
 - [ ] If methods have been deprecated, update rest of code to use the new methods

**Labels:**

 - [ ] If change to an IDD file, add the label `IDDChange`
 - [ ] If breaking existing API, add the label `APIChange`
 - [ ] If deemed ready, add label `Pull Request - Ready for CI` so that CI builds your PR

### Review Checklist

This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] Code Style, strip trailing whitespace, etc.
 - [ ] All related changes have been implemented: model changes, model tests, FT changes, FT tests, VersionTranslation, OS App
 - [ ] Labeling is ok
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified

### CODE REVIEW

File: resources/model/OpenStudio.idd [hunk 1]
```diff
@@ -28081,6 +28197,269 @@ OS:WaterHeater:HeatPump:WrappedCondenser,
...
+       \note Evaporator inlet air wet-bulb temperature corresponding to rated coil performance
+       \note (heating capacity, COP and SHR).
+  N5,  \field Rated Condenser Inlet Water Temperature
+       \type real
+       \units C
+       \minimum> 25
+       \required-field
+       \note Condenser inlet water temperature corresponding to rated coil performance
+       \note (heating capacity, COP and SHR).
+  N6,  \field Rated Evaporator Air Flow Rate
```
Comment (jmarrec) [inline resources/model/OpenStudio.idd:28247]: I'd have made this one and the other autocalculatable fields `\required-field`. But your API is correct at least

```c++
    boost::optional<double> ratedEvaporatorAirFlowRate() const;

    bool isRatedEvaporatorAirFlowRateAutocalculated() const;

    bool setRatedEvaporatorAirFlowRate(double ratedEvaporatorAirFlowRate);

    void autocalculateRatedEvaporatorAirFlowRate();
```

File: src/model/ScheduleTypeRegistry.cpp [hunk 2]
```diff
@@ -217,6 +217,7 @@ namespace model {
        OptionalDouble()},
       {"CoilHeatingWater", "Availability", "availabilitySchedule", false, "Availability", 0.0, 1.0},
       {"CoilHeatingWaterBaseboard", "Availability", "availabilitySchedule", false, "Availability", 0.0, 1.0},
+      {"CoilWaterHeatingAirToWaterHeatPumpVariableSpeed", "Grid Signal", "gridSignalSchedule", true, "", OptionalDouble(), OptionalDouble()},
```
Comment (jmarrec) [inline src/model/ScheduleTypeRegistry.cpp:220]: @joseph-robertson What is this? I can't find it anywhere in the IDD nor as a method.

File: src/model/test/AirLoopHVACUnitaryHeatPumpAirToAir_GTest.cpp [hunk 3]
```diff
@@ -182,3 +184,52 @@ TEST_F(ModelFixture, AirLoopHVACUnitaryHeatPumpAirToAir_VariableSpeedCoils) {
...
+  CoilHeatingDXVariableSpeed shdwhHeatingCoil(m);
+  CoilWaterHeatingAirToWaterHeatPumpVariableSpeed shdwhWaterHeatingCoil(m);
+
+  CoilSystemIntegratedHeatPumpAirSource coilSystem(m, spaceCoolingCoil, spaceHeatingCoil, dedicatedWaterHeatingCoil, scwhCoil, scdwhCoolingCoil,
+                                                   scdwhWaterHeatingCoil, shdwhHeatingCoil, shdwhWaterHeatingCoil);
+
+  CoilHeatingElectric coilHeatingElectric(m, s);
+
+  AirLoopHVACUnitaryHeatPumpAirToAir testObject(m, s, supplyFan, coilSystem, coilSystem, coilHeatingElectric);
+
```
Comment (jmarrec) [inline src/model/test/AirLoopHVACUnitaryHeatPumpAirToAir_GTest.cpp:208]: I haven't seen the coilSystem.containingHVACComponent() method tested neither in this file nor in the src/model/test/CoilSystemIntegratedHeatPumpAirSource_GTest.cpp ?

File: src/model/test/CoilWaterHeatingAirToWaterHeatPumpVariableSpeedSpeedData_GTest.cpp [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+TEST_F(ModelFixture, CoilWaterHeatingAirToWaterHeatPumpVariableSpeedSpeedData) {
+  ::testing::FLAGS_gtest_death_test_style = "threadsafe";
+
+  ASSERT_EXIT(
+    {
+      Model m;
+      CoilWaterHeatingAirToWaterHeatPumpVariableSpeedSpeedData coil(m);
+
+      exit(0);
```
Comment (jmarrec) [inline src/model/test/CoilWaterHeatingAirToWaterHeatPumpVariableSpeedSpeedData_GTest.cpp:46]: Add actual testing for this class please

File: src/model/test/CoilWaterHeatingAirToWaterHeatPumpVariableSpeed_GTest.cpp [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+
+  CoilWaterHeatingAirToWaterHeatPumpVariableSpeed coil(m);
+  CoilWaterHeatingAirToWaterHeatPumpVariableSpeedSpeedData speed1(m);
+  coil.addSpeed(speed1);
+  coil.remove();
+
+  auto curves = m.getModelObjects<model::Curve>();
+
+  EXPECT_EQ(count, m.modelObjects().size() - curves.size());
+}
```
Comment (jmarrec) [inline src/model/test/CoilWaterHeatingAirToWaterHeatPumpVariableSpeed_GTest.cpp:156]: Please also test what happens to the SpeedDataList and the SpeedDatas themselves when cloning / removing.

File: src/model/test/CoilWaterHeatingAirToWaterHeatPumpVariableSpeed_GTest.cpp [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+using namespace openstudio;
+using namespace openstudio::model;
+
+TEST_F(ModelFixture, CoilWaterHeatingAirToWaterHeatPumpVariableSpeed_CoilWaterHeatingAirToWaterHeatPumpVariableSpeed) {
+  ::testing::FLAGS_gtest_death_test_style = "threadsafe";
+
+  ASSERT_EXIT(
+    {
+      Model m;
+      CoilWaterHeatingAirToWaterHeatPumpVariableSpeed coil(m);
```
Comment (jmarrec) [inline src/model/test/CoilWaterHeatingAirToWaterHeatPumpVariableSpeed_GTest.cpp:48]: I would like to see tests about adding it to the types that accept it (which according to IDD should be both CoilSystemIntegrated AND directly onto HeatPumpWaterHeater(PumpedCondenser)) and the containingHVACComponent() stuff somewhere (either in parent or child).

**Review by jmarrec — CHANGES_REQUESTED** [review]
Almost perfect. Couple of minor changes requested (I'll make most of them and clean up my review comments) and a bit more testing would be nice.

File: src/energyplus/ForwardTranslator/ForwardTranslateCoilWaterHeatingAirToWaterHeatPumpVariableSpeed.cpp [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+  boost::optional<IdfObject>
+    ForwardTranslator::translateCoilWaterHeatingAirToWaterHeatPumpVariableSpeed(CoilWaterHeatingAirToWaterHeatPumpVariableSpeed& modelObject) {
+    IdfObject idfObject(IddObjectType::Coil_WaterHeating_AirToWaterHeatPump_VariableSpeed);
+    m_idfObjects.push_back(idfObject);
+
+    boost::optional<std::string> s;
+    boost::optional<double> value;
+
+    if (modelObject.speeds().size() == 0) {
+      // TODO: if it has zero speed data, Log and don't translate
```
[the remark this answers is not in the document]
Comment (jmarrec) [reply src/energyplus/ForwardTranslator/ForwardTranslateCoilWaterHeatingAirToWaterHeatPumpVariableSpeed.cpp:57]: Done 3e0150349d7b14c103df02ac44b88a2737d58448

File: src/model/CoilWaterHeatingAirToWaterHeatPumpVariableSpeed.cpp [hunk 8]
```diff
@@ -0,0 +1,727 @@
...
+      OS_ASSERT(other.iddObject().type() == CoilWaterHeatingAirToWaterHeatPumpVariableSpeed::iddObjectType());
+    }
+
+    CoilWaterHeatingAirToWaterHeatPumpVariableSpeed_Impl::CoilWaterHeatingAirToWaterHeatPumpVariableSpeed_Impl(
+      const CoilWaterHeatingAirToWaterHeatPumpVariableSpeed_Impl& other, Model_Impl* model, bool keepHandle)
+      : HVACComponent_Impl(other, model, keepHandle) {}
+
+    const std::vector<std::string>& CoilWaterHeatingAirToWaterHeatPumpVariableSpeed_Impl::outputVariableNames() const {
+      static const std::vector<std::string> result{
+        // TODO
```
[the remark this answers is not in the document]
Comment (jmarrec) [reply src/model/CoilWaterHeatingAirToWaterHeatPumpVariableSpeed.cpp:79]: Done 503a590336396c083b89399830f284ad0188b71f

File: src/model/CoilWaterHeatingAirToWaterHeatPumpVariableSpeed.cpp [hunk 9]
```diff
@@ -0,0 +1,727 @@
...
+      return CoilWaterHeatingAirToWaterHeatPumpVariableSpeed::iddObjectType();
+    }
+
+    std::vector<ScheduleTypeKey> CoilWaterHeatingAirToWaterHeatPumpVariableSpeed_Impl::getScheduleTypeKeys(const Schedule& schedule) const {
+      std::vector<ScheduleTypeKey> result;
+      UnsignedVector fieldIndices = getSourceIndices(schedule.handle());
+      UnsignedVector::const_iterator b(fieldIndices.begin()), e(fieldIndices.end());
+
+      return result;
+    }
```
[the remark this answers is not in the document]
Comment (jmarrec) [reply src/model/CoilWaterHeatingAirToWaterHeatPumpVariableSpeed.cpp]: Done 503a590336396c083b89399830f284ad0188b71f

File: src/model/CoilWaterHeatingAirToWaterHeatPumpVariableSpeed.cpp [hunk 10]
```diff
@@ -0,0 +1,727 @@
...
+      }
+    }
+
+    boost::optional<double> CoilWaterHeatingAirToWaterHeatPumpVariableSpeed_Impl::autocalculatedRatedEvaporatorAirFlowRate() const {
+      return getAutosizedValue("TODO", "W");
+    }
+
+    boost::optional<double> CoilWaterHeatingAirToWaterHeatPumpVariableSpeed_Impl::autocalculatedRatedCondenserWaterFlowRate() const {
+      return getAutosizedValue("TODO", "m3/s");
+    }
```
[the remark this answers is not in the document]
Comment (jmarrec) [reply src/model/CoilWaterHeatingAirToWaterHeatPumpVariableSpeed.cpp]: I added an actual "// TODO" explaining why. I tried to locate it in the source code of E+ but I don't think it actually reports it anywhere.

File: src/model/CoilSystemIntegratedHeatPumpAirSource.cpp [hunk 11]
```diff
@@ -0,0 +1,751 @@
...
+          }
+        } else if (boost::optional<HVACComponent> coil = airLoopHVACUnitaryHeatPumpAirToAir.heatingCoil()) {
+          if (coil->handle() == this->handle()) {
+            return airLoopHVACUnitaryHeatPumpAirToAir;
+          }
+        }
+      }
+
+      return boost::none;
+    }
```
[the remark this answers is not in the document]
Comment (jmarrec) [reply src/model/CoilSystemIntegratedHeatPumpAirSource.cpp:208]: | Type                                                  | Reference                               |
|-------------------------------------------------------|-----------------------------------------|
| OS:CoilSystem:IntegratedHeatPump:AirSource            | IntegratedHeatPumps                     |
| OS:Coil:WaterHeating:AirToWaterHeatPump:VariableSpeed | HeatPumpWaterHeaterDXCoilsVariableSpeed |


OS:WaterHeaterHeatPump, field DX Coil references either `IntegratedHeatPumps` or directly the `HeatPumpWaterHeaterDXCoilsVariableSpeed` (I don't know if that makes sense or not)


```
OS:WaterHeaterHeatPump
  [...]
  A12, \field DX Coil
       \required-field
       \type object-list
       \object-list HeatPumpWaterHeaterDXCoils
       \object-list HeatPumpWaterHeaterDXCoilsVariableSpeed
       \object-list IntegratedHeatPumps
```

[the remark this answers is not in the document]
Comment (jmarrec) [reply src/model/CoilSystemIntegratedHeatPumpAirSource.cpp:208]:
```ruby
m = Model.new
hpwh = WaterHeaterHeatPump.new(m)
ci = CoilSystemIntegratedHeatPumpAirSource.new(m)
hpwh.setDXCoil(ci)
=> true

ci.containingHVACComponent.empty?
=> true
```

File: src/model/test/CoilWaterHeatingAirToWaterHeatPumpVariableSpeedSpeedData_GTest.cpp (hunk 4 above)
Comment (jmarrec) [reply src/model/test/CoilWaterHeatingAirToWaterHeatPumpVariableSpeedSpeedData_GTest.cpp:46]: Done in 0091ee313fee5e1702b700eed04ffe1535581a0d

File: src/model/ScheduleTypeRegistry.cpp (hunk 2 above)
Comment (jmarrec) [reply src/model/ScheduleTypeRegistry.cpp:220]: Removed in 1ca23846362495dc84903eb821a16b0b645baec6

File: src/model/test/AirLoopHVACUnitaryHeatPumpAirToAir_GTest.cpp (hunk 3 above)
Comment (jmarrec) [reply src/model/test/AirLoopHVACUnitaryHeatPumpAirToAir_GTest.cpp:208]: DOne in 7c9bcdcbc98b2529d22c637207d746ec3ca5becd

File: src/model/CoilSystemIntegratedHeatPumpAirSource.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (jmarrec) [reply src/model/CoilSystemIntegratedHeatPumpAirSource.cpp:208]: Now supported in 12e1a52c08b88dda279dd6bfeb3db5e093a7521f and tested for in 7c9bcdcbc98b2529d22c637207d746ec3ca5becd

File: src/model/test/CoilWaterHeatingAirToWaterHeatPumpVariableSpeed_GTest.cpp (hunk 6 above)
Comment (jmarrec) [reply src/model/test/CoilWaterHeatingAirToWaterHeatPumpVariableSpeed_GTest.cpp:48]: Done in 501a283ae2c0eea888c040c0e749fb6c0a8a349a

File: src/model/test/CoilWaterHeatingAirToWaterHeatPumpVariableSpeed_GTest.cpp (hunk 5 above)
Comment (jmarrec) [reply src/model/test/CoilWaterHeatingAirToWaterHeatPumpVariableSpeed_GTest.cpp:158]: Done in f0409908eeb8aa66a715847f96a9b0f6f298e25c

**Review by jmarrec — APPROVED** [review]
@joseph-robertson I made all changes I wanted to see. Please review to make sure I didn't miss anything or did something stupid. Tested that the OS-res test still passes (https://github.com/NREL/OpenStudio-resources/pull/136)

**Comment by joseph-robertson** [comment]
@jmarrec, great review. And thanks for punching out all of the items, and for checking the test on OS-res.

**Pull request merged**